### PR TITLE
(tick-testing 6/6) Add dry run mutations for schedules

### PR DESCRIPTION
**Pull request by dpeng817**
Adds two new things to the gql API:
- A field on the schedules schema that retrieves a specified number of tick timestamps in the future / past from a specified timestamp.
- A mutation that performs a dry run of a schedule.

Test cases:
- Evaluate a typical schedule
- Attempt to evaluate a non-existent schedule
- Evaluate a schedule which fails in user code

### CODE REVIEW

**Comment by dpeng817** [comment]
Current dependencies on/for this PR:
* master
  * **PR #11869** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/11869" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/dagster-io/dagster/11869?utm_source=stack-comment).

File: python_modules/dagster/dagster/_core/host_representation/external.py [hunk 1]
```diff
@@ -617,9 +617,11 @@ def get_current_instigator_state(
                 ScheduleInstigatorData(self.cron_schedule, start_timestamp=None),
             )
 
-    def execution_time_iterator(self, start_timestamp: float) -> Iterator[datetime]:
+    def execution_time_iterator(
+        self, start_timestamp: float, is_forward: bool = True
```
Comment (prha) [inline python_modules/dagster/dagster/_core/host_representation/external.py:621]: consider ascending/descending?

File: python_modules/dagster-graphql/dagster_graphql/schema/schedules/schedules.py [hunk 2]
```diff
@@ -130,6 +134,26 @@ def resolve_futureTicks(self, _graphene_info, **kwargs):
...
+        tick_times = []
+        time_iter = self._external_schedule.execution_time_iterator(cursor, is_forward=False)
+
+        limit = kwargs.get("limit", 10)
+
+        for _ in range(limit):
+            tick_times.append(next(time_iter).timestamp())
+
+        past_ticks = [
+            GrapheneFutureInstigationTick(self._schedule_state, tick_time)
```
Comment (prha) [inline python_modules/dagster-graphql/dagster_graphql/schema/schedules/schedules.py:150]: Mostly disconcerted that we can return future ticks in the past.

some suggestions in random order of seriousness:

- UnfulfilledTick
- PotentialTick
- TickEvaluation
- UnrealizedTick
- ProtoTick
- TickPromise

**Review by prha — DISMISSED** [review]

File: python_modules/dagster-graphql/dagster_graphql/schema/schedules/schedules.py (hunk 2 above)
Comment (dpeng817) [reply python_modules/dagster-graphql/dagster_graphql/schema/schedules/schedules.py:150]: Went with PotentialTick: https://github.com/dagster-io/dagster/pull/11904

**Review by alangenfeld — COMMENTED** [review]
I'm pretty confused by whats going on here. I think "past ticks" is a very misleading name.

It looks like this is mostly just following the future tick pattern, but is this the best API for building out the front end product experience we want? Do you want to have to go forward/backward N ticks or should we facilitate selecting any time or any tick aligned with the cron schedule?

File: python_modules/dagster-graphql/dagster_graphql/schema/instigation.py [hunk 3]
```diff
@@ -177,13 +177,16 @@ class GrapheneDryRunInstigationTick(graphene.ObjectType):
...
+        self,
+        selector: Union[ScheduleSelector, SensorSelector],
+        timestamp: Optional[float],
+        cursor: Optional[str] = None,
+    ):
         self._selector = check.inst_param(selector, "selector", (ScheduleSelector, SensorSelector))
         self._timestamp = timestamp
         self._cursor = cursor
 
-    def resolve_timestamp(self, graphene_info: ResolveInfo):
```
Comment (salazarm) [inline python_modules/dagster-graphql/dagster_graphql/schema/instigation.py:185]: You removed this by accident and its causing the type generation to think the timestamp field doesnt exist
Screen Shot 2023-02-02 at 11 54 21 AM

Comment (dpeng817) [reply python_modules/dagster-graphql/dagster_graphql/schema/instigation.py:185]: hrm I didn't directly remove it from the gql layer so not sure what's happening here. Will investigate.

Comment (salazarm) [reply python_modules/dagster-graphql/dagster_graphql/schema/instigation.py]: removing `def resolve_timestamp` is akin to removing it from the gql layer

File: python_modules/dagster-graphql/dagster_graphql/schema/schedules/schedules.py [hunk 4]
```diff
@@ -41,6 +41,12 @@ class GrapheneSchedule(graphene.ObjectType):
     futureTick = graphene.NonNull(
         GrapheneDryRunInstigationTick, tick_timestamp=graphene.NonNull(graphene.Int)
     )
+    ticksFromTimestamp = graphene.NonNull(
+        graphene.List(graphene.Float),
+        start_timestamp=graphene.Float(),
+        upper_limit=graphene.Int(),
+        lower_limit=graphene.Int(),
+    )
```
Comment (alangenfeld) [inline python_modules/dagster-graphql/dagster_graphql/schema/schedules/schedules.py:49]: what other names did you consider besides `ticksFromTimestamp` ?

**Review by alangenfeld — COMMENTED** [review]
I think its worth a few more test cases for the args against `ticksFromTimestamp`

File: python_modules/dagster-graphql/dagster_graphql/schema/schedules/schedules.py (hunk 4 above)
Comment (dpeng817) [reply python_modules/dagster-graphql/dagster_graphql/schema/schedules/schedules.py:49]: another is potentialTickTimestamps, timestampsInRange

File: python_modules/dagster-graphql/dagster_graphql/schema/instigation.py (hunk 3 above)
Comment (dpeng817) [reply python_modules/dagster-graphql/dagster_graphql/schema/instigation.py]: Should be resolved now tho

File: python_modules/dagster-graphql/dagster_graphql_tests/graphql/test_scheduler.py [hunk 5]
```diff
@@ -437,6 +603,23 @@ def test_next_tick(graphql_context):
...
+        graphql_context,
+        GET_SCHEDULE_QUERY,
+        variables={"scheduleSelector": schedule_selector, "ticksAfter": cur_timestamp},
+    )
+
+    ticks = result.data["scheduleOrError"]["ticksFromTimestamp"]
+    assert len(ticks) == 6
+    assert len([tick for tick in ticks if tick > cur_timestamp]) == 3
+    assert len([tick for tick in ticks if tick <= cur_timestamp]) == 3
+
```
Comment (alangenfeld) [inline python_modules/dagster-graphql/dagster_graphql_tests/graphql/test_scheduler.py:621]: add testing to cover the implementation complexity you have comments for like preventing ticks in forward and backward iterators on exact match

File: python_modules/dagster-graphql/dagster_graphql/schema/schedules/schedules.py (hunk 4 above)
Comment (alangenfeld) [reply python_modules/dagster-graphql/dagster_graphql/schema/schedules/schedules.py:49]: I think i like `potentialTickTimestamps` most

**Review by alangenfeld — DISMISSED** [review]
See inline comments

File: python_modules/dagster-graphql/dagster_graphql_tests/graphql/test_scheduler.py (hunk 5 above)
Comment (dpeng817) [reply python_modules/dagster-graphql/dagster_graphql_tests/graphql/test_scheduler.py:621]: this test does cover the forward and backward iterator case - will add a comment denoting that though

**Review by alangenfeld — APPROVED** [review]

**Pull request merged**